### PR TITLE
Fixes: #4316 Added support for key-word based get()

### LINKED ISSUE #4316: keyword-based get() for composite primary key
**Migrated issue, originally created by Alex Rothberg**

Right now the `ident` parameter to `query.get` is a tuple in the case of composite primary keys with the stipulation:

>For a composite primary key,
         the order of identifiers corresponds in most cases
         to that of the mapped :class:`.Table` object's
         primary key columns.

Having to know and hard code the order of the primary keys seems less than ideal. I suggest accepting a dict mapping names to the values for the primary keys. My implementation of this looks like:

```python
from sqlalchemy.inspection import inspect

def get_by_pks(model, **kwargs):
    return model.query.get(
        tuple(
            kwargs[key.name]
            for key in inspect(model).primary_key
        )
    )
```

### PR DESCRIPTION
**Pull request by sanjana96**
This fixes #4316 
 
This pull request is:
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.

**Have a nice day!**

### CODE REVIEW

File: test/orm/test_query.py [hunk 1]
```diff
@@ -633,6 +633,42 @@ class Foo(object):
...
+
+        s = Session()
+        assert s.query(CompositePk).get({"i": 100, "j": 100}) is None
+
+    def test_get_composite_pk_keyword_based_result(self):
+        CompositePk = self.classes.CompositePk
+
+        s = Session()
+        one_two = s.query(CompositePk).get({"i": 1, "j": 2})
+        assert one_two.i == 1
```
Comment (zzzeek) [inline test/orm/test_query.py]: eq_(one_two.i, 1)

File: test/orm/test_query.py [hunk 2]
```diff
@@ -633,6 +633,42 @@ class Foo(object):
 
 
 class GetTest(QueryTest):
+    def test_get_composite_pk_keyword_based_no_result(self):
+        CompositePk = self.classes.CompositePk
+
+        s = Session()
+        assert s.query(CompositePk).get({"i": 100, "j": 100}) is None
```
Comment (zzzeek) [inline test/orm/test_query.py]: is_(s.query(CompositePK).get(....), None)

`is_()` is available from sqlalchemy.testing if it's not imported already

File: lib/sqlalchemy/orm/query.py [hunk 3]
```diff
@@ -991,17 +986,32 @@ def _get_impl(self, primary_key_identity, db_load_fn, identity_token=None):
         if hasattr(primary_key_identity, "__composite_values__"):
             primary_key_identity = primary_key_identity.__composite_values__()
 
-        primary_key_identity = util.to_list(primary_key_identity)
-
         mapper = self._only_full_mapper_zero("get")
 
+        is_dict = isinstance(primary_key_identity, dict)
```
Comment (zzzeek) [inline lib/sqlalchemy/orm/query.py]: very nice! :)

File: lib/sqlalchemy/orm/query.py [hunk 4]
```diff
@@ -991,17 +986,32 @@ def _get_impl(self, primary_key_identity, db_load_fn, identity_token=None):
...
         if len(primary_key_identity) != len(mapper.primary_key):
             raise sa_exc.InvalidRequestError(
                 "Incorrect number of values in identifier to formulate "
                 "primary key for query.get(); primary key columns are %s"
                 % ",".join("'%s'" % c for c in mapper.primary_key)
             )
 
+        if is_dict:
+            try:
+                primary_key_identity = list(
```
Comment (zzzeek) [inline lib/sqlalchemy/orm/query.py]: so this part, needs to be based on the mapped attributes:

   primary_key_identity[prop.key] for prop in mapper._identity_key_props

File: lib/sqlalchemy/orm/query.py [hunk 5]
```diff
@@ -991,17 +986,32 @@ def _get_impl(self, primary_key_identity, db_load_fn, identity_token=None):
...
+                                             primary_key_identity[value.name]
+                                             for value in mapper.primary_key
+                                        )
+            except KeyError:
+                raise sa_exc.InvalidRequestError(
+                    "Incorrect names of values in identifier to formulate "
+                    "primary key for query.get(); primary key columns are %s"
+                    % ",".join("'%s'" % c for c in mapper.primary_key)
```
Comment (zzzeek) [inline lib/sqlalchemy/orm/query.py]: `prop.key for prop in mapper._identity_key_props`

File: lib/sqlalchemy/orm/query.py [hunk 6]
```diff
@@ -991,17 +986,32 @@ def _get_impl(self, primary_key_identity, db_load_fn, identity_token=None):
...
+        if is_dict:
+            try:
+                primary_key_identity = list(
+                                             primary_key_identity[value.name]
+                                             for value in mapper.primary_key
+                                        )
+            except KeyError:
+                raise sa_exc.InvalidRequestError(
+                    "Incorrect names of values in identifier to formulate "
+                    "primary key for query.get(); primary key columns are %s"
```
Comment (zzzeek) [inline lib/sqlalchemy/orm/query.py]: "primary key attribute names are"

**Review by zzzeek — CHANGES_REQUESTED** [review]
looks great, some easy changes then we can run through CI.

**Comment by sanjana96** [comment]
@zzzeek Thanks for the review, I have made all the requested changes, hope this works!

**Review by zzzeek — APPROVED** [review]

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work to try to get revision a5cf95a132946e69288406fc74d455edbce8c848 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change a5cf95a132946e69288406fc74d455edbce8c848: https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1137

File: lib/sqlalchemy/orm/query.py [hunk 7]
```diff
@@ -920,12 +920,7 @@ def get(self, ident):
 
         :param ident: A scalar or tuple value representing
          the primary key.   For a composite primary key,
-         the order of identifiers corresponds in most cases
```
Comment (cancan101) [inline lib/sqlalchemy/orm/query.py]: should the doc string indicate that the ordered list is still supported (ie keep the old text and just append with the comment about a dict)?

Comment (zzzeek) [reply lib/sqlalchemy/orm/query.py]: yes definitely.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work to try to get revision cfa8297ad2490be9eae24ec8b1a691e43cd75868 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset cfa8297ad2490be9eae24ec8b1a691e43cd75868 added to existing Gerrit review https://gerrit.sqlalchemy.org/#/c/sqlalchemy/sqlalchemy/+/1137

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

Code-Review+2 Workflow+1

thanks Sanjana for all the effort on this!  looks great

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/1137 has been **merged**. Congratulations! :)

**Comment by sanjana96** [comment]
> **mike bayer** (zzzeek) wrote:
> 
> Code-Review+2 Workflow+1
> 
> thanks Sanjana for all the effort on this! looks great

My pleasure, thanks for being so patient with all the mistakes :D

**Comment by nurgasemetey** [comment]
Any example or docs?

**Pull request closed without merging**